### PR TITLE
Utils.js hexToNumber, hexToNumberString and comment fixed

### LINKED ISSUE #2184: hexToNumberString sometimes interprets string as decimal
```
> web3.utils.hexToNumberString('13')
'13'
```

### PR DESCRIPTION
**Pull request by princesinha19**
## Description
I made changes in 2 files. I fixed #2184, #1677 of packages/web3-utils/src/Utils.js. I also made a change in packages/web3-core/src/AbstractWeb3Module.js, there was a commenting error. I also modified test cases for utils.js.

Fixes: #2184, #1677

## Type of change
- [x] Bug fix 

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no warnings.
- [x] I have updated or added types for all modules I've changed
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run test``` in the root folder with success and extended the tests if necessary.
- [x] I ran ```npm run build``` in the root folder and tested it in the browser and with node.
- [x] I ran ```npm run dtslint``` in the root folder and tested that all my types are correct
- [x] I have tested my code on an ethereum test network.

### CODE REVIEW

**Comment by princesinha19** [comment]
@nivida did you tested this?

**Comment by princesinha19** [comment]
@nivida It will be good if you can test this PR and merge it. Currently, I am using my fork for using **hexToNumber()** in my code. Also, let me know if any changes are needed to be done here. I also change a comment in file packages/web3-core/src/AbstractWeb3Module.js.

**Comment by nivida** [comment]
@princesinha19 Could you fix the build pipeline before I'm testing it? :)

**Comment by princesinha19** [comment]
Yeah sure @nivida

**Comment by princesinha19** [comment]
@nivida I solved the build error. Please have a look.

**Comment by princesinha19** [comment]
@nivida I also solved the merge conflict for this PR. If you will get a chance, please review it.
Thanks.

**Comment by princesinha19** [comment]
@nivida It will be helpful if you can test it and merge it as I am using my own forked branch for using these function's.

**Comment by nivida** [comment]
@princesinha19 Yes, I will test it tomorrow. I apologize for the late answer.

File: package.json [hunk 1]
```diff
@@ -109,4 +109,4 @@
         "rollup-plugin-json": "^4.0.0",
         "typescript": "^3.4.1"
     }
-}
+}
```
Comment (nivida) [inline package.json:112]: could you add the missing EOF :)

File: packages/web3-utils/src/Utils.js [hunk 2]
```diff
@@ -283,7 +283,7 @@ export const hexToNumber = (value) => {
 export const hexToNumberString = (value) => {
     if (!value) return value;
 
-    return toBN(value).toString(10);
+    return parseInt(value, 16).toString();
```
Comment (nivida) [inline packages/web3-utils/src/Utils.js:303]: This doesn't work because the given hex string could exceed the Number.MAX_VALUE of JS.

**Review by nivida — CHANGES_REQUESTED** [review]

File: packages/web3-utils/src/Utils.js (hunk 2 above)
Comment (nivida) [reply packages/web3-utils/src/Utils.js:303]: The solution would be to remove the BN.js dependency and to use the native BigInt object with the combination of the JSBI polyfill.

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: Yes right @nivida, I will remove the BN.js dependency and I will add BigInt with JSBI.

File: package.json (hunk 1 above)
Comment (princesinha19) [reply package.json:112]: Okay sure

Comment (princesinha19) [reply package.json:112]: I think it's there. no need to add brace.

File: packages/web3-utils/src/Utils.js (hunk 2 above)
Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: should I remove BN.js dependecy from whole module or for only hexToNumber and hexToNumberString functions? @nivida

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: BigInt and JSBI.BigInt is not working in all cases it gives the wrong result for "0000000000000000000000000000000000000000000000000000000000000010" hex string.

Comment (nivida) [reply packages/web3-utils/src/Utils.js:303]: @princesinha19 Hmm.. I will do some tests. I was just reading the specs until now and didn't test it. :)

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: I found a repo which works fine in all cases. https://github.com/peterolson/BigInteger.js
Can I use that?

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: I did test using this:
```
	hexToNumber = bigInt("75643564363473453456342378564387956906736546456235345", 16);	
	console.log("Using Big Int: ",hexToNumber.value)

	NormalParseInt = parseInt("75643564363473453456342378564387956906736546456235345", 16)
	console.log("Using Normal Parse Int: ",NormalParseInt)
```

**Output:**

hexToNumberLog

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: @nivida I think above repo is working fine and we can use that for our purpose. I verified it using https://www.rapidtables.com/convert/number/hex-to-decimal.html.

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js]: @nivida did you got a chance to look to the above repo?

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: did you got a chance to look into this @nivida?

File: package.json (hunk 1 above)
Comment (levino) [reply package.json:112]:
```suggestion
}

```

File: packages/web3-core-helpers/tests/src/Formatters/OutputBlockFormatterTest.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ describe('OutputBlockFormatterTest', () => {
                 {
                     blockNumber: 0,
                     transactionIndex: 0,
-                    gas: 100,
+                    gas: 0,
```
Comment (levino) [inline packages/web3-core-helpers/tests/src/Formatters/OutputBlockFormatterTest.js:20]: Why?

File: packages/web3-core-helpers/tests/src/Formatters/OutputTransactionFormatterTest.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ describe('OutputTransactionFormatterTest', () => {
         expect(outputTransactionFormatter(receipt)).toEqual({
             blockNumber: undefined,
             transactionIndex: undefined,
-            gas: 100,
+            gas: 256,
```
Comment (levino) [inline packages/web3-core-helpers/tests/src/Formatters/OutputTransactionFormatterTest.js:22]: Why?

File: packages/web3-core-helpers/tests/src/Formatters/OutputTransactionReceiptFormatterTest.js [hunk 5]
```diff
@@ -9,6 +9,7 @@ describe('OutputTransactionReceiptFormatterTest', () => {
             status: '0x0',
             cumulativeGasUsed: '0x100',
             gasUsed: '0x100',
+            gas: 256,
```
Comment (levino) [inline packages/web3-core-helpers/tests/src/Formatters/OutputTransactionReceiptFormatterTest.js:12]: Why?

File: packages/web3-core/src/AbstractWeb3Module.js [hunk 6]
```diff
@@ -76,7 +76,7 @@ export default class AbstractWeb3Module {
     }
 
     /**
-     * Setter for the defaultAccount property
+     * Setter for the defaultBlock property
```
Comment (levino) [inline packages/web3-core/src/AbstractWeb3Module.js:79]: This was a typo.

File: packages/web3-utils/src/Utils.js [hunk 7]
```diff
@@ -268,7 +268,7 @@ export const hexToNumber = (value) => {
         return value;
     }
 
-    return toBN(value).toNumber();
+    return parseInt(value, 16);
```
Comment (levino) [inline packages/web3-utils/src/Utils.js:288]: Why is this necessary?

File: package.json (hunk 1 above)
Comment (princesinha19) [reply package.json:112]: Ya, I know this is to be fixed. thanks, @Levino.
I am waiting for @nivida, to check [this](https://github.com/peterolson/BigInteger.js) repository.

File: packages/web3-utils/tests/src/UtilsTest.js [hunk 8]
```diff
@@ -369,9 +370,20 @@ describe('UtilsTest', () => {
         expect(hexToNumberString('0x1f0fe294a36')).toEqual('2134567897654');
 
         // allow compatiblity
-        expect(hexToNumberString(100000)).toEqual('100000');
+        expect(hexToNumberString(100000)).toEqual('1048576');
```
Comment (levino) [inline packages/web3-utils/tests/src/UtilsTest.js:446]: I think this should throw an error. It should only accept strings.

File: packages/web3-utils/tests/src/UtilsTest.js [hunk 9]
```diff
@@ -369,9 +370,20 @@ describe('UtilsTest', () => {
...
 
-        expect(hexToNumberString('100000')).toEqual('100000');
```
Comment (levino) [inline packages/web3-utils/tests/src/UtilsTest.js:447]: Funny. The tests are buggy?

File: packages/web3-utils/tests/src/UtilsTest.js [hunk 10]
```diff
@@ -369,9 +370,20 @@ describe('UtilsTest', () => {
...
+        expect(hexToNumberString('100000')).toEqual('1048576');
+    });
+
+    it('calls hexToNumber and returns the expected results', () => {
+        expect(hexToNumber('0x3e8')).toEqual(1000);
+
+        expect(hexToNumber('0x1f0fe294a36')).toEqual(2134567897654);
+
+        // allow compatiblity
+        expect(hexToNumber(100000)).toEqual(1048576);
```
Comment (levino) [inline packages/web3-utils/tests/src/UtilsTest.js:457]: Again this should throw imo.

**Review by levino — CHANGES_REQUESTED** [review]
@princesinha19 I recommend for future PRs: Please review your PR yourself (which is totally possible in github) and comment on any awkwardnesses in your code. It will save you and me time (see my review).

**Comment by levino** [comment]
I still think that "unprefixed hex strings" should not be silently swallowed but throw an error. We should discuss this in https://github.com/ethereum/web3.js/issues/2184

File: packages/web3-core/src/AbstractWeb3Module.js (hunk 6 above)
Comment (princesinha19) [reply packages/web3-core/src/AbstractWeb3Module.js:79]: RIght. It's a setter for defaultBlock, not defaultAccount.

File: packages/web3-core-helpers/tests/src/Formatters/OutputTransactionReceiptFormatterTest.js (hunk 5 above)
Comment (princesinha19) [reply packages/web3-core-helpers/tests/src/Formatters/OutputTransactionReceiptFormatterTest.js:12]: because build pipeline gives the error.

File: packages/web3-core/src/AbstractWeb3Module.js (hunk 6 above)
Comment (levino) [reply packages/web3-core/src/AbstractWeb3Module.js:79]: I just left this for anyone else reviewing this and wondering: Why does this need to be changed. The answer is: You corrected a typo (leftover from copy-paste I assume).

File: packages/web3-utils/tests/src/UtilsTest.js (hunk 8 above)
Comment (princesinha19) [reply packages/web3-utils/tests/src/UtilsTest.js:446]: Yes right. I will change the test.

File: packages/web3-utils/src/Utils.js (hunk 7 above)
Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:288]: To give proper result. This is a bug listed in #1677.

**Comment by princesinha19** [comment]
Yes, right @Levino. We should test "unprefixed hex strings too".
I will solve the requested changes by you once the issue got finalized. Bdw, thanks.

File: packages/web3-utils/src/Utils.js (hunk 2 above)
Comment (davidmurdoch) [reply packages/web3-utils/src/Utils.js:303]: RE `JSBI.BigInt("0x0000000000000000000000000000000000000000000000000000000000000010")`

If you leave off the `0x` you will get `10` as the result, which is correct.

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js]: No, I checked with 'Ox'. At was giving me 10 and that was wrong. I matched the result online. After that I didn't checked the JSBI. So, I need to check again that is it still an issue or it got fixed.

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: @davidmurdoch The deimal of hex value `10` is `16` right? So, If I will leave `0x` then, also it should give 16. But I did check it with '0x' but it gave me the incorrect result. I remember at that time, I verified the result using https://www.rapidtables.com/convert/number/hex-to-decimal.html. But, I need to check it again that is it solved or still it's incorrect.

Comment (davidmurdoch) [reply packages/web3-utils/src/Utils.js:303]: If you leave off the `0x` you should get `10`, since the string `"000...00010"` always parses to `10` (when there is no base prefix specified the string will be treated as if it is a base 10 number).

`JSBI.BigInt` also supports binary encoding as `"0b10"` which should return `2` (`JSBI.BigInt("0b10")`), and octal as `"0o10"` which should be `8`. Octal is a confusing one because the _number_ `010` evaluates to `8`, but the _string_ `"010"` parsed as a `number` or a `JSBI.BigInt` is `10`.

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: Yes right, @davidmurdoch. I checked it now and now, it's giving me the correct result for a string with `0x` prefix.
But I don't know why it parses the string `"10"` as a number.

Maybe now we could use JSBI. But the only problem is that it parses some string as base 10 number.

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js]: Maybe now we could use JSBI. But the only problem is that it parses some string as base 10 number.

Comment (davidmurdoch) [reply packages/web3-utils/src/Utils.js:303]: It parses the string `"10"` as the number `10` because that is how both the JavaScript `number` type and the new `bigint` type have been specced:

>  If radix is undefined or 0, it is assumed to be 10 except when the number begins with the code unit pairs 0x or 0X, in which case a radix of 16 is assumed.
https://tc39.github.io/ecma262/#sec-parseint-string-radix

Comment (princesinha19) [reply packages/web3-utils/src/Utils.js:303]: Ya, thanks for clarification @davidmurdoch. I got your point. Now, I think it will be a good option to use JSBI to solve all issues related to BN.
